File: web/assets/js/expert-inspection.js
/**
 * Expert Inspection Module
 * Handles expert-level industrial quality inspection task form submission and result display
 */

let formState = {
    isLoading: false,
    currentResult: null,
    previewUrl: null,
};

/**
 * Initialize form on page load
 */
document.addEventListener('DOMContentLoaded', () => {
    console.log('Initializing expert inspection system...');
    
    setupFormListeners();
    autofillTaskId();
});

/**
 * Auto-fill task ID with unique value
 */
function autofillTaskId() {
    const taskIdInput = document.getElementById('task-id');
    if (taskIdInput && !taskIdInput.value) {
        taskIdInput.value = generateTaskId();
    }
}

/**
 * Setup form event listeners
 */
function setupFormListeners() {
    const form = document.getElementById('detection-form');
    const imageInput = document.getElementById('image-input');
    const dropZone = document.getElementById('drop-zone');

    // Form submission
    if (form) {
        form.addEventListener('submit', handleFormSubmit);
    }

    if (imageInput) {
        imageInput.addEventListener('change', () => {
            const file = imageInput.files?.[0] || null;
            updateImagePreview(file);
        });
    }

    // Drag and drop support
    if (dropZone && imageInput) {
        dropZone.addEventListener('click', () => imageInput.click());
        
        dropZone.addEventListener('dragover', (e) => {
            e.preventDefault();
            dropZone.style.borderColor = '#2563eb';
            dropZone.style.background = '#eff6ff';
        });

        dropZone.addEventListener('dragleave', () => {
            dropZone.style.borderColor = '#cbd5e1';
            dropZone.style.background = '#f8fafc';
        });

        dropZone.addEventListener('drop', (e) => {
            e.preventDefault();
            dropZone.style.borderColor = '#cbd5e1';
            dropZone.style.background = '#f8fafc';
            
            const file = e.dataTransfer.files?.[0];
            if (file) {
                imageInput.files = e.dataTransfer.files;
                updateImagePreview(file);
            }
        });
    }
}

function updateImagePreview(file) {
    const container = document.getElementById('image-preview-container');
    const previewImage = document.getElementById('image-preview');
    const previewMeta = document.getElementById('image-preview-meta');

    if (!container || !previewImage) return;

    if (!file) {
        clearImagePreview();
        return;
    }

    if (!file.type || !file.type.startsWith('image/')) {
        clearImagePreview();
        showError('请上传有效的图像文件');
        return;
    }

    if (formState.previewUrl) {
        window.URL.revokeObjectURL(formState.previewUrl);
    }

    const url = window.URL.createObjectURL(file);
    formState.previewUrl = url;
    previewImage.src = url;
    container.style.display = 'block';

    if (previewMeta) {
        previewMeta.textContent = `${file.name} · ${formatFileSize(file.size)}`;
    }
}

function clearImagePreview() {
    const container = document.getElementById('image-preview-container');
    const previewImage = document.getElementById('image-preview');
    if (container) container.style.display = 'none';
    if (previewImage) previewImage.src = '';
    if (formState.previewUrl) {
        window.URL.revokeObjectURL(formState.previewUrl);
        formState.previewUrl = null;
    }
}

function formatFileSize(bytes) {
    if (bytes === 0) return '0 B';
    const k = 1024;
    const sizes = ['B', 'KB', 'MB', 'GB'];
    const i = Math.floor(Math.log(bytes) / Math.log(k));
    return parseFloat((bytes / Math.pow(k, i)).toFixed(2)) + ' ' + sizes[i];
}

/**
 * Handle form submission
 */
async function handleFormSubmit(e) {
    e.preventDefault();

    const taskId = document.getElementById('task-id').value.trim();
    const assetId = document.getElementById('asset-id').value.trim();
    const category = document.getElementById('category-input').value;
    const imageFile = document.getElementById('image-input')?.files?.[0];

    if (!assetId || !imageFile) {
        showError('请完善资产编号和待检图像');
        return;
    }

    showLoading(true);

    try {
        const formData = new FormData();
        formData.append('task_id', taskId);
        formData.append('asset_id', assetId);
        formData.append('category', category);
        formData.append('start_time', new Date().toISOString());
        formData.append('end_time', new Date().toISOString());
        formData.append('data_source', 'expert_inspection');
        formData.append('question', document.getElementById('question-input')?.value?.trim() || '');
        formData.append('image', imageFile);
        
        const parameters = {
            tool_type: 'qwen3.5-plus',
            category: category
        };
        formData.append('parameters', JSON.stringify(parameters));

        console.log('开始专家质检任务...', taskId);
        const result = await apiClient.runDetection(formData);

        formState.currentResult = result;
        displayResult(result);
        
    } catch (error) {
        console.error('专家质检失败:', error);
        showError(error.message || '质检执行过程中发生错误');
    } finally {
        showLoading(false);
    }
}

/**
 * Display expert detection results
 */
function displayResult(result) {
    const resultsSection = document.getElementById('results-section');
    if (!resultsSection) return;

    // Set headers
    document.getElementById('result-task-id-header').textContent = `检测结论 - ${result.task_id}`;
    
    // Status Badge
    const statusBadge = document.getElementById('status-badge');
    const isAnomaly = result.status.toLowerCase() === 'anomaly';
    statusBadge.textContent = isAnomaly ? '发现异常 (Anomaly)' : '正常 (Normal)';
    statusBadge.className = `status-badge ${isAnomaly ? 'status-anomaly' : 'status-normal'}`;

    // Thought Chain
    document.getElementById('expert-thought').textContent = result.thought || '暂无推理过程。';

    // Detailed Findings (Explanation)
    const exp = result.explanation || {};
    document.getElementById('visual-evidence').textContent = exp.visual_evidence || '未提取到明显的视觉差异证据。';
    document.getElementById('standard-reference').textContent = exp.standard_reference || '基于通用工业标准。';
    document.getElementById('root-cause').textContent = exp.root_cause_analysis || '尚不明确，建议进一步排查。';
    
    // Confidence Score
    let scoreText = 'N/A';
    if (result.anomalies && result.anomalies.length > 0) {
        const maxScore = Math.max(...result.anomalies.map(a => a.score || 0));
        scoreText = (maxScore * 100).toFixed(1) + '%';
    } else if (isAnomaly) {
        scoreText = '高置信度';
    } else {
        scoreText = '100.0%';
    }
    document.getElementById('confidence-score').textContent = scoreText;

    // Action Recommendation
    document.getElementById('action-recommendation').textContent = exp.action_recommendation || '暂无明确建议。';

    // Markdown Summary
    const summaryEl = document.getElementById('report-summary');
    if (typeof renderMarkdownLite === 'function') {
        summaryEl.innerHTML = renderMarkdownLite(result.summary || '未生成报告摘要。');
    } else {
        summaryEl.textContent = result.summary || '未生成报告摘要。';
    }

    resultsSection.style.display = 'block';
    resultsSection.scrollIntoView({ behavior: 'smooth' });
}

function showLoading(show) {
    formState.isLoading = show;
    const overlay = document.getElementById('loading-overlay');
    const submitBtn = document.querySelector('button[type="submit"]');

    if (overlay) overlay.style.display = show ? 'flex' : 'none';
    if (submitBtn) submitBtn.disabled = show;
}

function showError(message) {
    const errorDiv = document.getElementById('error-message');
    if (errorDiv) {
        errorDiv.textContent = message;
        errorDiv.style.display = 'block';
        setTimeout(() => { errorDiv.style.display = 'none'; }, 5000);
    }
}

function resetForm() {
    const form = document.getElementById('detection-form');
    if (form) form.reset();
    autofillTaskId();
    clearImagePreview();
    document.getElementById('results-section').style.display = 'none';
}

function downloadReport() {
    if (!formState.currentResult) return;
    const data = JSON.stringify(formState.currentResult, null, 2);
    const blob = new Blob([data], { type: 'application/json' });
    const url = URL.createObjectURL(blob);
    const a = document.createElement('a');
    a.href = url;
    a.download = `expert-report-${formState.currentResult.task_id}.json`;
    a.click();
    URL.revokeObjectURL(url);
}

// Reuse helper from api-client.js or define here if missing
if (typeof generateTaskId !== 'function') {
    window.generateTaskId = () => `TASK-${Date.now()}-${Math.random().toString(36).substr(2, 5).toUpperCase()}`;
}

File: app/prompts/expert_inspection.py
from __future__ import annotations

from langchain_core.prompts import ChatPromptTemplate

EXPERT_INSPECTION_PROMPT = ChatPromptTemplate.from_messages(
    [
        (
            "system",
            (
                "[系统指令]\n"
                "你是一位资深工业质量检测专家（Industrial Inspection Agent），负责依据生产标准对零件进行多维分析。"
                "你擅长执行“观察-对比-校验-决策”的闭环推理流程，并严格依据证据生成结论。\n\n"
                "[上下文信息]\n"
                "待检对象类别: {object_category}\n"
                "检索库 A (历史案例/图片库): {library_a}\n"
                "检索库 B (领域知识库/规程库): {library_b}\n\n"
                "[任务要求]\n"
                "请遵循 MMAD (Multimodal Industrial Anomaly Detection) 测评协议，分析待检图像。"
                "你需要输出详细的推理链（thought），并给出最终判定结果。\n\n"
                "[证据对比 - In-Context Comparison]:\n"
                "1. 将待检图与[检索库A]中的“正常参考图”进行像素级/结构级差异点对齐。\n"
                "2. 确认异常区域是“物理破损（划痕、污染）”还是“逻辑偏差（位置偏移、漏装）”。\n\n"
                "[置信度自评 - Confidence Evaluation]:\n"
                "评估本次判断的确定性（0.0-1.0）。\n"
                "评估准则：视觉特征是否清晰？检索到的案例与当前样本的相似度（μ）是否足够高？逻辑推理是否自洽（σ）？\n"
                "计算得分：S = μ - λ σ (λ 为罚分系数，通常取 0.2)。\n\n"
                "[输出格式]\n"
                "必须以 JSON 格式输出，包含以下字段：\n"
                "{{\n"
                '  "thought": "描述完整的观察、找茬对比、规程核对以及置信度计算过程。",\n'
                '  "result": {{\n'
                '    "status": "Normal / Anomaly",\n'
                '    "defect_type": "缺陷类别（如无则填 None）",\n'
                '    "location": "缺陷在图像中的坐标或方位描述",\n'
                '    "confidence_score": 0.95\n'
                '  }},\n'
                '  "explanation": {{\n'
                '    "visual_evidence": "详细的视觉差异描述",\n'
                '    "standard_reference": "引用了规程库中的哪一条标准",\n'
                '    "root_cause_analysis": "基于知识库分析的可能失效原因",\n'
                '    "action_recommendation": "给生产线的处置建议"\n'
                '  }}\n'
                "}}\n\n"
                "注意：输出必须仅包含合法 JSON，不要有任何多余的 Markdown 标记或文本。"
            ),
        ),
        (
            "user",
            (
                "任务ID: {task_id}\n"
                "资产ID: {asset_id}\n"
                "待检测异常详情(JSON): {anomalies}\n"
                "用户补充问题: {question}\n"
            ),
        ),
    ]
)
